core/perf_test/PerfTest_Stream.cpp

```diff
@@ -0,0 +1,297 @@
+  const size_t N8                 = std::pow(state.range(0), 8);
+  static constexpr int DATA_RATIO = 2;
+
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> a("a", N8);
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> b("b", N8);
+
+  perform_set(a, aInit);
+
+  for (auto _ : state) {
+    Kokkos::Timer timer;
+    perform_copy(a, b);
+    KokkosBenchmark::report_results(state, a, DATA_RATIO, timer.seconds());
+  }
+
+  if (validate_array(b, aInit)) {
+    state.SkipWithError("validation failure");
+  }
+}
+
+template <unsigned MemTraits>
+static void StreamScale(benchmark::State& state) {
+  const size_t N8                 = std::pow(state.range(0), 8);
+  static constexpr int DATA_RATIO = 2;
+
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> a("a", N8);
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> b("b", N8);
+
+  perform_set(b, bInit);
+
+  for (auto _ : state) {
+    Kokkos::Timer timer;
+    perform_scale(a, b, scalar);
+    KokkosBenchmark::report_results(state, b, DATA_RATIO, timer.seconds());
+  }
+
+  if (validate_array(a, bInit * scalar)) {
+    state.SkipWithError("validation failure");
+  }
+}
+
+template <unsigned MemTraits>
+static void StreamAdd(benchmark::State& state) {
+  const size_t N8                 = std::pow(state.range(0), 8);
+  static constexpr int DATA_RATIO = 3;
+
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> a("a", N8);
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> b("b", N8);
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> c("c", N8);
+
+  perform_set(a, aInit);
+  perform_set(b, bInit);
+  perform_set(c, cInit);
+
+  for (auto _ : state) {
+    Kokkos::Timer timer;
+    perform_add(a, b, c);
+    KokkosBenchmark::report_results(state, c, DATA_RATIO, timer.seconds());
+  }
+
+  if (validate_array(c, aInit + bInit)) {
+    state.SkipWithError("validation failure");
+  }
+}
+
+template <unsigned MemTraits>
+static void StreamTriad(benchmark::State& state) {
+  const size_t N8                 = std::pow(state.range(0), 8);
+  static constexpr int DATA_RATIO = 3;
+
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> a("a", N8);
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> b("b", N8);
+  Kokkos::View<double*, Kokkos::MemoryTraits<MemTraits>> c("c", N8);
+
+  perform_set(a, aInit);
+  perform_set(b, bInit);
+  perform_set(c, cInit);
+
+  for (auto _ : state) {
+    Kokkos::Timer timer;
+    perform_triad(a, b, c, scalar);
+    KokkosBenchmark::report_results(state, a, DATA_RATIO, timer.seconds());
+  }
+
+  if (validate_array(a, bInit + scalar * cInit)) {
+    state.SkipWithError("validation failure");
+  }
+}
+
+// skips a benchmark with an error from thrown exceptions
+template <void (*bm)(benchmark::State&)>
+static void or_skip(benchmark::State& state) {
+  try {
+    bm(state);
+  } catch (const std::runtime_error& e) {
+    state.SkipWithError(e.what());
+  }
+};
+
+namespace Test {
+
+BENCHMARK(or_skip<StreamSet<0>>)
```

Comment:
There is quite a bit of repetition here in writing the list of benchmarks. Perhaps writing a helper macro could be a good solution to avoid the duplication.